hack/push-cip.sh

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env bash
+
+# Copyright 2020 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# About: This script builds and pushes the cip and auditor images to a public repository,
+# defined in workspace_status.sh.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+repo_root=$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && pwd -P)
+# Detect which container variant to build.
+build_variant="cip"
+
+printUsage() {
+    >&2 echo "Usage: $0 [--audit]"
+}
+
+if (( $# > 1 )) || { (( $# > 0 )) && [[ "$1" != --audit ]];}; then
+    >&2 echo "ERROR: Invalid arguments."
+    printUsage
+    exit 1
+elif (( $# == 1 )) && [[ "$1" == --audit ]]; then
+    build_variant="auditor"
+fi
+
+# Takes array of images and pushes to their tagged registry location.
+pushImages() {
+    local images="${*}"
+    for img in $images; do
+        docker push "$img"
+    done
+}
+
+# Builds either auditor or cip container image from repo_root. Also adds
+# tags, given by an array of tags.
+buildImage() {
+    # Add build variant argument.
+    local variant=$1
+    local cmd="docker build --build-arg variant=$variant "
+    # Concatenate tags.
+    shift
+    local tags="${*}"
+    for tag in $tags; do
+        cmd+="-t $tag "
+    done
+    # Specify Dockerfile location.
+    cmd+="$repo_root"
+    # Build the container.
+    >&2 echo "Executing: $cmd"
+    $cmd
+}
+
+# Builds and pushes the image variant with specified tags.
+shipImage() {
+    local variant=$1
+    shift
+    buildImage "$variant" "${*}"
+    pushImages "${*}"
+}
+
+# Entrypoint of this script.
+main() {
+    # Inject workspace variables.
+    source <("${repo_root}"/workspace_status.sh inject)
+
+    # Change the repo for cloudbuild jobs only.
+    if [[ ! -z "${CLOUDBUILD_REPO+x}" ]]; then
+        STABLE_IMG_REPOSITORY=${CLOUDBUILD_REPO}
+    fi
+
+    # Assemble image tag prefixes.
+    local stable_tag_prefix=${STABLE_IMG_REGISTRY}/${STABLE_IMG_REPOSITORY}/${STABLE_IMG_NAME}
+    local test_tag_prefix=${STABLE_TEST_AUDIT_STAGING_IMG_REPOSITORY}/${STABLE_IMG_NAME}
+
+    # Define tags for the auditor.
+    # NOTE: Both cip and auditor variants build an auditor image. Although they are named differently,
```

Comment:
Hmm if the content is exactly the same, we technically don't have to name the two differently. The `stable_tag_prefix` vs the `test_tag_prefix` distinction is already good enough I think.

In summary there are 2 possibilities:

```
(1) build the `-auditor` and cip images:
    "${stable_tag_prefix}-auditor:latest"
    "${stable_tag_prefix}-auditor:${STABLE_IMG_TAG}"
    "${stable_tag_prefix}:latest"
    "${stable_tag_prefix}:${STABLE_IMG_TAG}"
(2) build the -auditor-test image only:
    "${test_tag_prefix}-auditor-test:latest"
    "${test_tag_prefix}-auditor-test:${STABLE_IMG_TAG}"
```

So, why not just have 1 if/else branch? Also in the future we could remove the `-auditor` vs `-auditor-test` naming convention because it's unnecessary as stated above.

Reply:
We don't have 1 if/else branch because that design decision would be less 'DRY'. Since both variants MUST build an auditor, there should only be one such statement to do so: `shipImage "auditor" $tags[@]`. If we split things up into 1 if/else branch, it would look something like this:
```
    if [[ "$build_variant" == auditor ]]; then
        audit_tags=(
            "${test_tag_prefix}-auditor-test:latest"
            "${test_tag_prefix}-auditor-test:${STABLE_IMG_TAG}"
        )
        # Build and push image auditor.
        shipImage "auditor" "${audit_tags[@]}"
    else
        local audit_tags=(
            "${stable_tag_prefix}-auditor:latest"
            "${stable_tag_prefix}-auditor:${STABLE_IMG_TAG}"
        )
        # Build and push image cip.
        shipImage "auditor" "${audit_tags[@]}"

        local cip_tags=(
            "${stable_tag_prefix}:latest"
            "${stable_tag_prefix}:${STABLE_IMG_TAG}"
        )
        # Build and push image cip.
        shipImage "cip" "${cip_tags[@]}"
    fi
```
For this design, 
#### Pros:
- Branching is easy to identify
- Supporting more variants in the future is straightforward

#### Con:
- Repetitive auditor definitions

Following an earlier suggestion to be more "DRY" in code, I prioritized this in my decision-making. Personally, if I feel DRY applies for 3+ repetitions. SInce this is such a simple script, I would be more than happy to prioritize readability.

Reply:
Another variant of implementation could be this:
```
    local audit_tags
    local cip_tags

    if [[ "$build_variant" == auditor ]]; then
        # Only build and push the auditor image.
        audit_tags=(
            "${test_tag_prefix}-auditor-test:latest"
            "${test_tag_prefix}-auditor-test:${STABLE_IMG_TAG}"
        )
    else
        # Build and push the auditor and cip image.
        audit_tags=(
            "${stable_tag_prefix}-auditor:latest"
            "${stable_tag_prefix}-auditor:${STABLE_IMG_TAG}"
        )
        cip_tags=(
            "${stable_tag_prefix}:latest"
            "${stable_tag_prefix}:${STABLE_IMG_TAG}"
        )
        shipImage "cip" "${cip_tags[@]}"
    fi

    shipImage "auditor" "${audit_tags[@]}"
```
But this actually appears as if I skip the auditor shipping, and is less clear. Preferring the first implementation.

Reply:
SG. You could even avoid the creation of the tag arrays and just do, e.g.,

```
shipImage "cip" \
    "${stable_tag_prefix}:latest" \
    "${stable_tag_prefix}:${STABLE_IMG_TAG}"
```

Because I don't think it hurts readability, and plus you don't really reuse these array variables anywhere else.

Reply:
Sure! Resolving.